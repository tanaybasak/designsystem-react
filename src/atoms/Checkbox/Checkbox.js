import React, { useState, useEffect } from 'react';
import PropTypes from 'prop-types';
import Tooltip from '../Tooltip';
import prefix from '../../settings';

export default function Checkbox({
  className,
  checked,
  label,
  indeterminate,
  tooltipTitle,
  tooltipDirection,
  ...restProps
}) {
  const [isChecked, setChecked] = useState(checked || false);
  const classnames = `${prefix}-checkbox-item ${className}`.trim();

  useEffect(() => {
    setChecked(checked);
  }, [checked]);

  return (
    <div className={classnames}>
      <input
        className={`${prefix}-checkbox`}
        type="checkbox"
        checked={isChecked}
        ref={el => el && (el.indeterminate = indeterminate)}
        {...restProps}
        onChange={event => {
          setChecked(!isChecked);
          if (restProps.onChange) {
            restProps.onChange(event);
          }
        }}
      />
      {tooltipTitle ? (
        <Tooltip content={tooltipTitle} direction={tooltipDirection}>
          <label className={`${prefix}-checkbox-label`} htmlFor={restProps.id}>
            {label}
          </label>
        </Tooltip>
      ) : (
        <label className={`${prefix}-checkbox-label`} htmlFor={restProps.id}>
          {label}
        </label>
      )}
    </div>
  );
}

Checkbox.propTypes = {
  /** Custom class on the Checkbox wrapper. */
  className: PropTypes.string,
  /** Disable Checkbox */
  disabled: PropTypes.bool,
  /** Text for Checkbox Label. */
  label: PropTypes.string,
  /** Tooltip Text for Checkbox Label. */
  tooltipTitle: PropTypes.string,
  /** Tooltip Direction eg: top, bottom, left, right */
  tooltipDirection: PropTypes.oneOf(['top', 'bottom', 'left', 'right']),
  /** Unique string for the Checkbox. */
  id: PropTypes.string.isRequired,
  /** indeterminate state for Checkbox */
  indeterminate: PropTypes.bool,
  /** Accepts event handler as prop/argument.
   *
   * @signature
   * ```event```: sends a callback event
   */
  onChange: PropTypes.func,
  /** Control Checked state for Checkbox */
  checked: PropTypes.bool,
  /** Unique identifier used within forms. */
  name: PropTypes.string,
  /** Text value to be considered when submitting forms. */
  value: PropTypes.oneOfType([PropTypes.string, PropTypes.number])
};

Checkbox.defaultProps = {
  className: '',
  disabled: false,
  label: '',
  tooltipTitle: null,
  tooltipDirection: 'bottom',
  onChange: () => {},
  checked: false,
  indeterminate: false
};
