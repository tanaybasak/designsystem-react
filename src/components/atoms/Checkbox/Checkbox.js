import React, { useState } from 'react';
import PropTypes from 'prop-types';
import { prefix } from '../../../settings';

export default function Checkbox({ className, checked, labelText, ...restProps }) {
    const [isChecked, setChecked] = useState(checked || false);
    const classnames = `${prefix}-checkbox-item ${className}`.trim();

    return (
        <div className={classnames}>
            <input
                className={`${prefix}-checkbox`}
                type="checkbox"
                checked={isChecked}
                aria-checked={isChecked}
                aria-disabled={restProps['disabled'] ? true : false}
                {...restProps}
                onChange={event => {
                    setChecked(!isChecked);
                    if (restProps.onChange) {
                        restProps.onChange(event);
                    }
                }}
            />
            {
                labelText ?
                    <label
                        className={`${prefix}-checkbox-label`}
                        htmlFor={restProps.id}
                    >
                        {labelText}
                    </label>
                    : null
            }
        </div>
    );
};

Checkbox.propTypes = {
    className: PropTypes.string,
    disabled: PropTypes.bool,
    labelText: PropTypes.string,
    id: PropTypes.string,
    onChange: PropTypes.func,
    checked: PropTypes.bool,
    title: PropTypes.string
};

Checkbox.defaultProps = {
    className: '',
    labelText: '',
    disabled: false,
    onChange: () => { },
    checked: false
};