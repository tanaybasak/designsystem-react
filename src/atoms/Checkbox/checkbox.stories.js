import React from 'react';
import { storiesOf } from '@storybook/react';
import { action } from '@storybook/addon-actions';
import { text, boolean, select } from '@storybook/addon-knobs';
//@update-path-build-start
import Checkbox from './Checkbox';
//@update-path-build-end
const directionOption = {
  Top: 'top',
  Bottom: 'bottom',
  Left: 'left',
  Right: 'right'
};
storiesOf('Components/Checkbox', module)
  .add(
    'default',
    () => (
      <Checkbox
        disabled={boolean('Disabled', false)}
        id="checkbox1"
        label={text('Label', 'Checkbox Label')}
        onChange={action('Checkbox-OnChange')}
        value="check1"
      />
    ),
    {
      info: {
        text: `Description About Checkbox Component \n

      import { Checkbox } from '@patron/patron-react/checkbox';`
      }
    }
  )
  .add(
    'indeterminate',
    () => (
      <Checkbox
        disabled={boolean('Disabled', false)}
        id="checkbox1"
        indeterminate={boolean('Indeterminate', true)}
        checked={boolean('Checked', false)}
        label={text('Label', 'Checkbox Label')}
        onChange={action('Checkbox-OnChange')}
        value="check1"
      />
    ),
    {
      info: {
        text: `Description About Checkbox Component \n
  
      import { Checkbox } from '@patron/patron-react/checkbox';`
      }
    }
  )
  .add(
    'horizontal',
    () => (
      <div className="hcl-checkbox-group">
        <Checkbox
          disabled={boolean('Disabled', false)}
          id="checkbox1"
          label={text('Label 1', 'Checkbox Label 1')}
          onChange={action('Checkbox-OnChange')}
          value="check1"
        />
        <Checkbox
          disabled={boolean('Disabled', false)}
          id="checkbox2"
          label={text('Label 2', 'Checkbox Label 2')}
          checked
          onChange={action('Checkbox-OnChange')}
          value="check2"
        />
        <Checkbox
          id="checkbox3"
          label={text('Label 3', 'Checkbox Label 3')}
          value="check3"
          disabled
        />
      </div>
    ),
    {
      info: {
        text: `Description About Checkbox Component \n
  
        import { Checkbox } from '@patron/patron-react/checkbox';`
      }
    }
  )
  .add(
    'vertical',
    () => (
      <div className="hcl-checkbox-group hcl-stack-vertical">
        <Checkbox
          disabled={boolean('Disabled', false)}
          id="checkbox1"
          label={text('Label 1', 'Checkbox Label 1')}
          onChange={action('Checkbox-OnChange')}
          value="check1"
        />
        <Checkbox
          disabled={boolean('Disabled', false)}
          id="checkbox2"
          label={text('Label 2', 'Checkbox Label 2')}
          checked
          onChange={action('Checkbox-OnChange')}
          value="check2"
        />
        <Checkbox
          id="checkbox3"
          label={text('Label 3', 'Checkbox Label 3')}
          disabled
          value="check3"
        />
      </div>
    ),
    {
      info: {
        text: `Description About Checkbox Component \n
  
        import { Checkbox } from '@patron/patron-react/checkbox';`
      }
    }
  )
  .add(
    'with tooltip',
    () => (
      <Checkbox
        disabled={boolean('Disabled', false)}
        id="checkbox1"
        label={text('Label', 'Checkbox Label')}
        tooltipTitle={text(
          'Tooltip Title',
          'Breif Definition of the dotted underlined word.'
        )}
        tooltipDirection={select(
          'Tooltip Direction',
          directionOption,
          'bottom'
        )}
        onChange={action('Checkbox-OnChange')}
        value="check1"
      />
    ),
    {
      info: {
        text: `Description About Checkbox Component \n

      import { Checkbox } from '@patron/patron-react/checkbox';`
      }
    }
  );
