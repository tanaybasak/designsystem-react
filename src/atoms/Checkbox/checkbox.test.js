import React from 'react';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';
import prefix from '../../settings';
import Checkbox from './Checkbox';

describe('<Checkbox> component', () => {
    it('snapshots/renders Checkbox correctly', () => {
        const tree = renderer.create(<Checkbox />).toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('renders checkobx with label', () => {
        const checkbox = mount((<Checkbox labelText='defaultcheckbox' />));
        expect(checkbox.text()).toBe('defaultcheckbox');
    });

    it('renders checkobx with default classname', () => {
        const checkbox = mount((<Checkbox />));
        expect(checkbox.find(`.${prefix}-checkbox`).hasClass('hcl-checkbox')).toBe(true);
    });

    it('renders checkbox with default unchecked', () => {
        const checkbox = mount((<Checkbox />));
        expect(checkbox.props().checked).toBeFalsy();
    });

    it('renders checkbox with default checked', () => {
        const checkbox = mount((<Checkbox checked />));
        expect(checkbox.props().checked).toBeTruthy();
    });

    // it('on state change, trigger onChange event ', () => {
    //     const mockcallback = jest.fn();
    //     const checkbox = mount((<Checkbox checked onChange={mockcallback} />));
    //     checkbox.simulate('onchange');
    //     expect(mockcallback.mock.calls.length).toEqual(1);
    // });
});