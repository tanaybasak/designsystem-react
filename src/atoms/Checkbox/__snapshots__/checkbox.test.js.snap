// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Checkbox> component snapshots/renders Checkbox correctly 1`] = `
<div
  className="hcl-checkbox-item"
>
  <input
    checked={false}
    className="hcl-checkbox"
    disabled={false}
    id="hellocheckbox"
    onChange={[Function]}
    onKeyDown={[Function]}
    type="checkbox"
  />
  <label
    className="hcl-checkbox-label"
    htmlFor="hellocheckbox"
  >
    
  </label>
</div>
`;
